### PR TITLE
[DTensor] Fix loss_parallel log_softmax backward for non-cross-entropy consumers

### LINKED ISSUE #190746: [DTensor] loss_parallel() gives identity gradients to any log_softmax not consumed by cross_entropy
## Bug description

Inside `with loss_parallel():`, `torch.distributed.tensor.parallel.loss.py` registers a process-wide custom handler for `aten._log_softmax_backward_data` that returns `grad_output` unchanged:

```python
def _log_softmax_backward_handler(op_call, args, kwargs):
    grad_output = cast(DTensor, args[0])
    input_dtype = cast(torch.dtype, args[3])
    return grad_output.to(input_dtype)
```

This is correct only when the `log_softmax` feeds `nll_loss`, because `_nll_loss_and_log_softmax_backward` fuses both backward steps and recomputes the softmax-Jacobian term (`return (grad_input + torch.exp(x)) * grad_output`), then relies on the log_softmax backward being a no-op. The handler has no consumer check, no graph check, and no opt-out. Any other consumer of a vocab-sharded `log_softmax` inside the context gets `d/dx log_softmax(x) = I` instead of `I - softmax(x) 1^T`, so every such gradient is wrong by `-softmax(x) * grad_output.sum(class_dim)`, silently. The forward handler computes a fully correct all-reduced `log_softmax`, so the forward looks right and only the backward is wrong.

## Repro (4 GPUs)

```python
# torchrun --standalone --nproc_per_node=4 repro.py
import torch, torch.nn.functional as F
from torch.distributed.tensor import distribute_tensor, Shard, init_device_mesh
from torch.distributed.tensor.parallel import loss_parallel

mesh = init_device_mesh("cuda", (4,))
x = torch.randn(4, 8, 512, device="cuda")   # class dim = -1, sharded
c = torch.randn(4, 8, 512, device="cuda")   # any downstream weighting
dx = distribute_tensor(x, mesh, [Shard(2)]).requires_grad_(True)
dc = distribute_tensor(c, mesh, [Shard(2)])

with loss_parallel():                        # a log_softmax NOT feeding cross_entropy
    (F.log_softmax(dx, dim=-1) * dc).sum().backward()

g = dx.grad.full_tensor().double()
p = torch.softmax(x.double(), -1)
ref = c.double() - p * c.double().sum(-1, keepdim=True)
print("grad == grad_output (identity, the bug):", torch.equal(dx.grad.full_tensor(), c))
print("max err vs true grad:", (g - ref).abs().max().item())
```

The returned gradient is `grad_output` bitwise, every element; `max err` is large. Verified failing on torch 2.9.1+cu128 and a 2.12 nightly, all seeds, fp32 and bf16, on 4 GPUs (also reproduced on a fresh 4x L40S run). The handler is unchanged since loss parallel landed (#119877, 2024), so every release from 2.3 is affected.

## Reachability

Protected: a `log_softmax` on a replicated DTensor, or sharded on a non-class dim, raises loudly from `_find_all_reduce_mesh_dim` rather than corrupting; plain (non-DTensor) tensors never reach this dispatcher.

Not protected: any code that enables TP loss parallelism and computes a second `log_softmax` over the sharded class dim inside the context, or wraps the whole backward in the context while such a `log_softmax` exists in the graph. The DPO/ORPO log-probability shape is the single most natural non-cross-entropy use of a vocab-sharded log_softmax. A public trainer already does exactly this: `ShinoharaHare/LLM-Training` (`src/llm_training/lms/dpo/dpo.py`, `.../orpo/orpo.py`) computes `logits.log_softmax(2)` inside `loss_parallel()` and wraps `super().backward(loss)` in the context, with no `nll_loss` in the graph, so every DPO/ORPO gradient under TP is silently wrong. NVIDIA NeMo `speechlm2` wraps its whole model backward in `loss_parallel()` too (correct today because it only uses cross_entropy, but one added auxiliary log_softmax over the vocab would corrupt).

Note: torchtitan main no longer uses `loss_parallel()` (it moved to an explicit local-tensor autograd Function), so the flagship framework is not a current victim; older pinned torchtitan revisions that wrapped forward+loss+backward in the context are exposed.

## Suggested fix

Two options, both closing the silent path:

1. Make it correct for all consumers: compute the true distributed `log_softmax` backward in `_log_softmax_backward_handler` (`grad_output - softmax(output) * all_reduce(grad_output.sum(class_dim))`) and drop the fused Jacobian term from `_nll_loss_and_log_softmax_backward` so the two compose. Cost: one extra all-reduce on the cross_entropy backward that the current fusion deliberately avoids.

2. Fail loud: raise from the backward handler when the `log_softmax` gradient does not come from the fused `nll_loss` path, turning a silent wrong-grad into an error. Cheaper, but reliably distinguishing the fused case from a standalone `log_softmax` backward at op-handler granularity is the open design question (a naive thread-local flag can false-positive across multi-loss graphs).

The docstring ("Currently only the cross-entropy loss is supported") is the only existing signal, and it is weak: it reads as a restriction on which loss you write, the forward handler affirmatively computes a correct distributed bare log_softmax (and PyTorch tests that forward), and the violation is silent.

## Prior art

No existing issue or PR covers the no-op backward for non-nll consumers. Related loss_parallel issues (#128636, #130549/#130550, #133032, #135178, #152016, #180936) address other aspects. Present on main as of the 2.12 nightly.

*This report was authored with assistance from Claude; the repro, mechanism, and reachability were human-reviewed.*

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx

### PR DESCRIPTION
**Pull request by PannenetsF**
## Problem

loss_parallel() in torch/distributed/tensor/parallel/loss.py registers a process-wide no-op _log_softmax_backward_handler that returns grad_output unchanged (identity). This is only correct for the fused cross-entropy path, where _nll_loss_and_log_softmax_backward computes the full gradient (grad_input + exp(x)) * grad_output (including the log_softmax Jacobian) and the no-op just passes it through.

Any other consumer inside loss_parallel() — e.g. DPO/ORPO's (log_softmax(logits) * weights).sum() which has no nll_loss in the graph — silently gets the identity gradient instead of the correct (I - softmax(x) * 1^T) Jacobian.

## Fix

1. _log_softmax_backward_handler now computes the true distributed log_softmax Jacobian:
   grad_input = grad_output - softmax(output) * all_reduce(grad_output.sum(dim, keepdim=True))
   (one all_reduce(SUM) across the TP mesh dim, since the class dim is sharded).

2. _nll_loss_and_log_softmax_backward now returns only grad_input * grad_output (the nll_loss contribution), dropping the fused + torch.exp(x) term. The log_softmax Jacobian is applied separately by the backward handler — same result for cross-entropy, correct for all consumers.

Cost: one extra all_reduce on the cross-entropy backward (a small batch-sized tensor).

## Validation (4x H20 GPUs)

- Repro: standalone log_softmax gradient is not the identity (bug fixed), max err vs true grad = 3.17e-07. Cross-entropy backward correct for none/mean/sum reductions (err < 1e-7).
- Regression tests: all 18 test_loss_parallel* tests pass (251s, exit code 0), including the new test_loss_parallel_standalone_log_softmax_backward and test_loss_parallel_cross_entropy_still_correct.

## Tests

- test_loss_parallel_standalone_log_softmax_backward (new)
- test_loss_parallel_cross_entropy_still_correct (new)
- Updated comm count assertion 0 -> 1 for log_softmax backward all_reduce.

Fixes #190746

### CODE REVIEW

**Comment by RohitRathore1** [comment]
hey, i was already working on it. You should respect the comment

**Comment by PannenetsF** [comment]
just missed the message. closed
